### PR TITLE
RSDK-2617 - move ESP32 encoder to component implementation

### PR DESCRIPTION
**Pull request by gvaradarajan**
Also includes the following

- implements a PWM/DIR Esp32 Motor
- implements Stop on gRPC level for motor (non-breaking change, calls SetPower with 0. To be changed in a future PR)

### CODE REVIEW

File: examples/esp32/esp32.rs [hunk 1]
```diff
@@ -57,7 +57,7 @@ fn main() -> anyhow::Result<()> {
         use micro_rdk::esp32::analog::Esp32AnalogReader;
         use micro_rdk::esp32::base::Esp32WheelBase;
         use micro_rdk::esp32::board::EspBoard;
-        use micro_rdk::esp32::motor::MotorEsp32;
+        use micro_rdk::esp32::motor::ABMotorEsp32;
```
Comment (randhid) [inline examples/esp32/esp32.rs:60]: Will these namings be considered breaking in the future?

File: src/esp32/pin.rs [hunk 2]
```diff
@@ -1,13 +1,10 @@
-use esp_idf_hal::gpio::{OutputMode, Pin, PinDriver};
+use esp_idf_hal::gpio::{Pin, PinDriver};
```
Comment (randhid) [inline src/esp32/pin.rs:1]: Why are we removing output modes of pins here? Seems like it's more complete than what we have in rdk

**Review by randhid — APPROVED** [review]
Have Nico look at the common files when he's back in this pull request and make it a priority to change anything he wants to be changed. If the encoder works I'm happy with you moving forward with single line.

File: examples/esp32/esp32.rs (hunk 1 above)
Comment (gvaradarajan) [reply examples/esp32/esp32.rs:60]: Once we have config consumption fully implemented and rely on model names like in RDK, they won't be

File: src/esp32/encoder.rs [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+        let pcnt = Box::new(PulseStorage {
+            acc: Arc::new(AtomicI32::new(0)),
+            unit: 0,
+        });
+        let mut enc = Esp32Encoder {
+            pulse_counter: pcnt,
+            config: pcnt_config_t {
+                pulse_gpio_num: a.pin(),
+                ctrl_gpio_num: b.pin(),
+                pos_mode: pcnt_count_inc,
```
Comment (npmenard) [inline src/esp32/encoder.rs:73]: let's create a ticket to make it so the config is passed as an argument should make a new struct with sane default for that too

File: src/esp32/encoder.rs [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+        }
+        let tot = self.pulse_counter.acc.load(Ordering::Relaxed) * 100 + i32::from(ctr);
+        Ok(tot)
+    }
+    pub fn setup_pcnt(&mut self) -> anyhow::Result<()> {
+        unsafe {
+            match esp_idf_sys::pcnt_unit_config(&self.config as *const pcnt_config_t) {
+                ESP_OK | ESP_ERR_INVALID_STATE => {}
```
Comment (npmenard) [inline src/esp32/encoder.rs:172]: remove ESP_ERR_INVALID_STATE it's a mistake from my side it shouldn't be considered as a success

**Review by npmenard — CHANGES_REQUESTED** [review]

**Comment by npmenard** [comment]
Done with first pass

File: src/esp32/encoder.rs [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+}
+
+pub struct Esp32Encoder<A, B> {
+    pulse_counter: Box<PulseStorage>,
+    config: pcnt_config_t,
+    a: A,
+    b: B,
+}
+
+impl<A, B> Esp32Encoder<A, B>
```
[the remark this answers is not in the document]
Comment (gvaradarajan) [reply src/esp32/encoder.rs:58]: by drop method, do you mean implementing the drop trait?

File: src/esp32/motor.rs [hunk 6]
```diff
@@ -245,32 +78,42 @@ where
...
-            max_duty
-        );
         self.pwm
             .set_duty(((max_duty as f64) * pct.abs()).floor() as u32);
         Ok(())
     }
     fn get_position(&mut self) -> anyhow::Result<i32> {
-        self.enc.position()
+        let props = self.enc.get_properties();
+        let pos_type = match props {
```
[the remark this answers is not in the document]
Comment (gvaradarajan) [reply src/esp32/motor.rs:87]: well I need to pass PositionType to the encoder, should I default to passing TICKS?

File: src/esp32/encoder.rs [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+        self.reset()
+    }
+}
+
+impl<A, B> Status for Esp32Encoder<A, B>
+where
+    A: InputPin + PinExt,
+    B: InputPin + PinExt,
+{
+    fn get_status(&mut self) -> anyhow::Result<Option<prost_types::Struct>> {
```
[the remark this answers is not in the document]
Comment (randhid) [reply src/esp32/encoder.rs]: No - board and motor have statuses, what's polled for encoder is GetPosition

File: src/esp32/motor.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (randhid) [reply src/esp32/motor.rs:87]: In go unspecified is okay - and then the driver can return what the position type is as an output.

File: src/esp32/encoder.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (npmenard) [reply src/esp32/encoder.rs:58]: yes

**Review by npmenard — APPROVED** [review]
LGTM

**Pull request merged**